addon/-private/state-machine/-base.ts

```diff
@@ -47,24 +56,126 @@ export default abstract class StateMachine extends EmberObject {
   }
 
   addStep(this: StateMachine, name: string) {
-    if (this.stepTransitions.includes(name)) {
+    if (this.stepsToAdd.includes(name)) {
       return;
     }
 
-    // Set the first step, if it hasn't been yet
-    if (!get(this, 'firstStep')) {
-      set(this, 'firstStep', name);
+    this.stepsToAdd.push(name);
+    return bind(
+      this,
+      scheduleOnce,
+      'afterRender',
+      this,
+      this.flushAdditionQueue
+    )();
+  }
+
+  removeStep(this: StateMachine, name: string) {
+    if (this.stepsToRemove.includes(name)) {
+      return;
     }
 
-    this.stepTransitions.pushObject(name);
+    this.stepsToRemove.push(name);
+    return bind(this, scheduleOnce, 'render', this, this.flushRemoveQueue)();
+  }
+
+  flushAdditionQueue(this: StateMachine) {
+    let { firstStep, currentStep, stepsToAdd, stepTransitions } = getProperties(
+      this,
+      'firstStep',
+      'currentStep',
+      'stepsToAdd',
+      'stepTransitions'
+    );
+    let lastStep;
+
+    A(stepsToAdd)
+      // don't add steps that are already included
+      .filter(name => !this.stepTransitions.includes(name))
+      .forEach(name => {
+        // Set the first step, if it hasn't been yet
+        if (!firstStep) {
+          firstStep = name;
+        }
+
+        this.stepTransitions.pushObject(name);
+
+        // Set the current step, if it hasn't been yet
+        if (!currentStep) {
+          currentStep = name;
+        }
+
+        // Set the last step to this new one
+        lastStep = name;
+      });
+
+    setProperties(this, {
+      firstStep,
+      currentStep,
+      lastStep,
+      stepsToAdd: A()
+    });
+  }
+
+  flushRemoveQueue(this: StateMachine) {
+    let {
+      firstStep,
+      currentStep,
+      lastStep,
+      stepsToRemove,
+      stepsToAdd,
+      stepTransitions
+    } = getProperties(
+      this,
+      'firstStep',
+      'currentStep',
+      'lastStep',
+      'stepsToRemove',
+      'stepsToAdd',
+      'stepTransitions'
+    );
 
-    // Set the current step, if it hasn't been yet
-    if (!get(this, 'currentStep')) {
-      set(this, 'currentStep', name);
+    //don't remove steps that are to be added
```

Comment:
If we're choosing not to remove steps that are queued to be added, should they by removed from the "to be added" queue?

Reply:
I don't understand. The line following the comment prevents the code from removing all of the names in the state transition every time data is changed. Every time the array of steps changes values, all of the steps are destroyed and then added back. So this check prevents losing which step is the current step.